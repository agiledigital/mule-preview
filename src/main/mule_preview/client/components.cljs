(ns mule-preview.client.components
  "The react components that render the Mule preview"
  (:require
   [reagent.core :as r]
   [react-dom :as react-dom]
   [clojure.string :refer [split replace]]
   [lambdaisland.uri :refer [join]]
   ["react-popper" :refer [Manager Reference Popper]])
  (:require-macros [mule-preview.client.macros :as m]))

(def default-component-mapping {:image "UnknownNode-48x32.png"})
(def default-category-image "org.mule.tooling.ui.modules.core.miscellaneous.large.png")

(defn- pluralise
  "Pluralise a given string value"
  [string]
  (str string "s"))

(defn map-kv [m f]
  (reduce-kv #(assoc %1 (f %2) %3) {} m))

(defn- normalise-name [name]
  (let [split-name (split name #":")
        first-component (first split-name)]
    (if (> (count split-name) 1)
      (str first-component ":*")
      first-component)))

(defn- name-to-category-url [name default-image mappings]
  (let [element-to-icon-map (:mappings mappings)
        mapping (get element-to-icon-map (keyword name))
        category (:category mapping)
        filename (if (some? category) (str category ".large.png") default-image)]
    (if-not (nil? filename)
      (str "img/icons/" filename)
      nil)))

(defn- name-to-img-url [name is-nested default-value mappings]
  (let [element-to-icon-map (:mappings mappings)
        mapping (get element-to-icon-map (keyword name) default-value)
        regular-filename (:image mapping)
        filename (if is-nested (get mapping :nested-image regular-filename) regular-filename)]
    (if-not (nil? filename)
      (str "img/icons/" filename)
      nil)))

(defn- labels-to-diff-icon-url [labels]
  (cond
    (:added labels) "img/plus.svg"
    (:edited labels) "img/is-not-equal-to.svg"
    (:removed labels) "img/minus.svg"
    :else nil))

(defn- name-to-css-class [name]
  (when name (let [normalized-name (replace name #":" "_")]
               (str "mule-" normalized-name))))

(defn- image
  ([url content-root] (image url "" content-root))
  ([url class content-root]
   (if-not (nil? url)
     [:img {:src (str (join content-root url)) :class class}]
     nil)))

(defn- child-container [children]
  (into [] (concat [:div {:class "container-children"}] children)))

(defn- arrow [content-root]
  (image "img/arrow-right-2x.png" "flow-arrow" content-root))

(defn delta-text [name [previous current]]
  (cond
    (and (some? previous) (some? current)) [:div
                                            {:key name} (str name ": ")
                                            [:span {:class ["previous"]} previous]
                                            " > "
                                            [:span {:class ["current"]} current]]
    (and (some? previous) (nil? current)) [:div
                                           {:key name}
                                           [:span {:class ["previous"]}
                                            (str  name ": ")
                                            previous
                                            " (Attribute Removed)"]]
    (and (nil? previous) (some? current)) [:div
                                           {:key name}
                                           [:span {:class ["current"]}
                                            (str  name ": ")
                                            current
                                            " (Attribute Added)"]]))

(defn tooltip-item [{:keys [name delta]}]
  (if (#{"content-hash"} name)
    [:div
     {:key name}
     "Content Changed"]
    (delta-text name delta)))

(defn tooltip-edited [change-record]
  [:div (->>
         change-record
         (remove #(#{"hash" "description"} (:name %)))
         (map tooltip-item))])


(defn tooltip-added []
  [:div "Element Added"])

(defn tooltip-removed []
  [:div "Element Removed"])

(defn tooltip [change-record labels location placement]
  [:div {:class ["mp-popover" placement]}
   [:h3 {:class "mp-popover-title"}
    (str "Line " (:line location) ", Column " (:column location))]
   [:div {:class "mp-popover-content"}
    (cond
      (:added labels) (tooltip-added)
      (:removed labels) (tooltip-removed)
      (:edited labels) (tooltip-edited change-record)
      :else nil)]])

(defn popper [change-record labels location showing-atom anchor-el]
  (when @showing-atom (react-dom/createPortal
                       (r/as-element [:div {:class "mp-popover-root"}
                                      [:> Popper {:placement "auto" :reference-element @anchor-el}
                                       (fn [props]
                                         (let [{:keys [ref style placement]} (js->clj props :keywordize-keys true)]
                                           (r/as-element [:div {:ref ref :style style :data-placement placement :class placement}
                                                          (tooltip change-record labels location placement)])))]])
                       (.-body js/document))))

(defn mule-component-inner [{:keys [name description css-class content-root location change-record showing-atom labels mappings]}]
  (let [anchor-el (clojure.core/atom nil)
        img-url (name-to-img-url name false default-component-mapping mappings)
        diff-icon-url (labels-to-diff-icon-url labels)
        category-url (name-to-category-url name default-category-image mappings)
        should-show-tooltip (or change-record (:added labels) (:removed labels))]
    (fn []
      [:div {:ref #(reset! anchor-el %)}
       [:div {:class ["component-container" css-class]
              :on-mouse-over (m/handler-fn (reset! showing-atom should-show-tooltip))
              :on-mouse-out  (m/handler-fn (reset! showing-atom false))}
        (when diff-icon-url (image diff-icon-url "diff-icon" content-root))
        [:div
         {:class ["component" name]}
         (image category-url "category-frame" content-root)
         (image img-url "icon" content-root)
         [:div {:class "label"} description]]]
       (popper change-record labels location showing-atom anchor-el)])))

(defn mule-component [props]
  (let [showing-atom (r/atom false)]
    (fn [] (mule-component-inner (assoc props :showing-atom showing-atom)))))

(defn mule-container-inner [{:keys [name description children css-class content-root location change-record showing-atom labels mappings]}]
  (let [anchor-el (clojure.core/atom nil)
        generated-css-class (name-to-css-class name)
        img-url (name-to-img-url name (some? children) nil mappings)
        category-url (name-to-category-url name default-category-image mappings)
        interposed-children (interpose (arrow content-root) children)
        child-container-component (child-container interposed-children)
        should-show-tooltip (or change-record (:added labels) (:removed labels))]
    (fn []
      [:div {:ref #(reset! anchor-el %)}
       [:div {:class ["container" generated-css-class css-class]
              :on-mouse-over (m/handler-fn (reset! showing-atom should-show-tooltip))
              :on-mouse-out  (m/handler-fn (reset! showing-atom false))}
        [:div {:class "container-title"} description]
        [:div {:class "container-inner"}
         [:div {:class "icon-container"}
          (image category-url "category-frame" content-root)
          (image img-url "icon container-image" content-root)]
         child-container-component]]
       (popper change-record labels location showing-atom anchor-el)])))

(defn mule-container [props]
  (let [showing-atom (r/atom false)]
    (fn [] (mule-container-inner (assoc props :showing-atom showing-atom)))))

; Exports for testing with Jest
(def ^:export MuleComponent (r/reactify-component mule-component-inner))
(def ^:export MuleContainer (r/reactify-component mule-container-inner))