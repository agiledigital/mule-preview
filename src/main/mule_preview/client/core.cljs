(ns mule-preview.client.core
  "The entry point to the web application"
  (:require
   [reagent.core :as r]
   [mule-preview.client.views.preview :refer [start-preview start-preview-url]]
   [mule-preview.client.views.diff :refer [start-diff start-diff-url]]))

(defn view [root-component]
  [:div {:class "mp root-component"} @root-component])

(defn mount-root [element root-component]
  (r/render [(partial view root-component)] element))

(defn ^:export mount-url-diff-on-element [element file-a file-b content-root]
  (let [root-component (r/atom [:div])]
    (start-diff-url file-a
                    file-b
                    root-component
                    content-root)
    (mount-root element root-component)))

(defn ^:export mount-url-preview-on-element [element file content-root]
  (let [root-component (r/atom [:div])]
    (start-preview-url
     file
     root-component
     content-root)
    (mount-root element root-component)))

(defn ^:export mount-diff-on-element [element content-a content-b content-root]
  (let [root-component (r/atom [:div])]
    (start-diff content-a
                content-b
                root-component
                content-root)
    (mount-root element root-component)))

(defn ^:export mount-preview-on-element [element content content-root]
  (let [root-component (r/atom [:div])]
    (start-preview
     content
     root-component
     content-root)
    (mount-root element root-component)))

(defn ^:dev/after-load init! []
  (mount-url-diff-on-element
   (.getElementById js/document "app")
   "/example_xml/nice-example.xml"
   "/example_xml/nice-example-diff.xml"
   "."))
