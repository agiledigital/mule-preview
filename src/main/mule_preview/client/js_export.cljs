(ns mule-preview.client.js-export
  "Entry point for dependant Javascript projects.
   Exports the function names with camel case to fit Javascript conventions"
  (:require
   [mule-preview.client.core :refer [mount-url-diff-on-element
                                     mount-url-preview-on-element
                                     mount-diff-on-element
                                     mount-preview-on-element]]))

(def ^:export mountUrlDiffOnElement mount-url-diff-on-element)
(def ^:export mountUrlPreviewOnElement mount-url-preview-on-element)
(def ^:export mountDiffOnElement mount-diff-on-element)
(def ^:export mountPreviewOnElement mount-preview-on-element)

