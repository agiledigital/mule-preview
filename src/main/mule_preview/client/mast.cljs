(ns mule-preview.client.mast
  "Functions to convert Mule XML data structures to an intermediate MAST structure"
  (:require
   [clojure.walk :refer [prewalk]]
   [clojure.set :refer [union]]
   [clojure.string :refer [starts-with? split replace capitalize]]
   [mule-preview.client.utils :refer [remove-location]]
   [mule-preview.client.mappings :refer [root-container horizontal-container-list
                                         vertical-container-list error-handler-component-list
                                         error-handler-container-list munit-container-list]]))

(defn- get-tag [node]
  (name (node :tag)))

(defn- get-munit-title [tag-name]
  (let [munit-flow-name (last (split tag-name #":"))
        spaced (replace munit-flow-name #"-" " ")]
    (capitalize spaced)))

(defn- is-error-handler [node]
  (let [tag (get-tag node)
        result (contains? error-handler-component-list tag)]
    result))

(defn- is-munit-mock-component [node]
  (let [tag (get-tag node)
        result (starts-with? tag "mock:")]
    result))

(defn- get-description [node]
  (let [doc-name (get-in node [:attributes :doc:name])
        name (get-in node [:attributes :name])]
    (or doc-name name)))

(defn- create-mule-component [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        attributes (:attributes node)]
    {:type :component
     :tag-name tag-name
     :description (or description tag-name)
     :hash (hash (:attributes node))
     :content-hash (hash (remove-location content))
     :labels labels
     :location (:location node)
     :attributes attributes}))

(defn- create-mule-container-component [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        attributes (:attributes node)]
    {:type :container
     :tag-name tag-name
     :description description
     :hash (hash (:attributes node))
     :content content
     :labels labels
     :location (:location node)
     :attributes attributes}))

(defn-
  create-mule-psuedo-container
  ([content]
   (create-mule-psuedo-container content #{} ""))
  ([content extra-labels]
   (create-mule-psuedo-container content extra-labels ""))
  ([content extra-labels title]
   {:type :container
    :tag-name "psuedo"
    :description title
    :content content
    :labels (union #{:horizontal} extra-labels)}))

(defn- process-error-container [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        {error-handlers true regular-components false}
        (group-by is-error-handler content)
        attributes (:attributes node)]
    {:type :error-container
     :tag-name tag-name
     :description description
     :attributes attributes
     :content [(create-mule-psuedo-container regular-components #{:top})
               (create-mule-psuedo-container error-handlers #{:bottom})]
     :labels labels
     :location (:location node)}))

(defn- process-munit-container [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        {mocks true regular-components false}
        (group-by is-munit-mock-component content)
        title (get-munit-title tag-name)
        is-split-flow (= tag-name "munit:test")
        attributes (:attributes node)]
    {:type :munit-container
     :tag-name tag-name
     :description description
     :attributes attributes
     :content [(create-mule-psuedo-container mocks #{:top} (if is-split-flow "Setup" title))
               (create-mule-psuedo-container regular-components #{:bottom} (if is-split-flow "Test" ""))]
     :labels labels
     :location (:location node)}))

(defn- transform-tag [node]
  (let [tag-name (get-tag node)
        is-root-container (= tag-name root-container)
        is-munit-container (contains? munit-container-list tag-name)
        is-error-handler-container (contains? error-handler-container-list tag-name)
        is-error-handler-component (contains? error-handler-component-list tag-name)
        is-horizontal-container (contains? horizontal-container-list tag-name)
        is-vertical-container (contains? vertical-container-list tag-name)]
    (cond
      is-root-container (create-mule-container-component node tag-name #{:root :vertical})
      is-munit-container (process-munit-container node tag-name #{:vertical :munit})
      is-error-handler-container (process-error-container node tag-name #{:vertical})
      is-error-handler-component (create-mule-container-component node tag-name #{:error-handler})
      is-horizontal-container (create-mule-container-component node tag-name #{:horizontal})
      is-vertical-container (create-mule-container-component node tag-name #{:vertical})
      :else  (create-mule-component node tag-name #{}))))

(defn- transform-fn [node]
  (if (contains? node :tag)
    (let [transformed-tag (transform-tag node)]
      transformed-tag)
    node))

(defn xml->mast [xml]
  (prewalk transform-fn xml))
