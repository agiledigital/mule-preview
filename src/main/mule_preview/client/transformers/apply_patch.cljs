(ns mule-preview.client.transformers.apply_patch
  "Functions that take the output of diffDOM and transform a MAST data structure to reflect the changes in the diff."
  (:require
   [clojure.walk :refer [prewalk]]
   [mule-preview.client.utils :refer [insert vec-remove map-values]]
   [mule-preview.client.diff-algorithms.diff-dom :refer [dom->mast]]
   [mule-preview.client.mappings :refer [root-container horizontal-container-list
                                         vertical-container-list error-handler-component-list
                                         error-handler-container-list]]))

(defn- prepare-path [path]
  (let [keyworded-path (map #(if (string? %) (keyword %) %) path)
        interposed (interpose :content keyworded-path)
        vector-path (vec interposed)]
    (vec (conj interposed :content))))

(defn- insert-removal [element-route original-element removal-map]
  (let [index (last element-route)
        parent (drop-last element-route)]
    (update removal-map parent #(conj % [index original-element]))))

(defn- shift-element [insert-index removed-index]
  (if (< insert-index removed-index) (inc removed-index) removed-index))

(defn- shift-removal [route removal-map]
  (let [index (last route)
        parent (drop-last route)]
    (update removal-map parent
            (fn [coll] (map-values #(shift-element index %) coll)))))

(defn- add-element [mast removal-map route element]
  (let [index (last route)
        keyword-route (drop-last (prepare-path route))
        original-element (dom->mast element)
        updated (update-in original-element [:labels] #(set (conj % :added)))]
    [(update-in mast keyword-route #(insert % index updated)) (shift-removal route removal-map)]))

(defn- modify-element [mast removal-map route element name oldValue newValue]
  (let [index (last route)
        keyword-route (prepare-path route)
        original-element (get-in mast keyword-route)
        with-labels (update-in original-element [:labels] #(set (conj % :edited)))
        with-description (assoc-in with-labels [(keyword name)] newValue)
        with-change-record (update-in with-description [:change-record]
                                      #(conj % {:name name :delta [oldValue newValue]}))]
    [(assoc-in mast keyword-route with-change-record) removal-map]))

(defn- remove-element [mast removal-map route]
  (let [index (last route)
        element-route (prepare-path route)
        keyword-route (drop-last element-route)
        original-element (get-in mast element-route)]
    [(update-in mast keyword-route #(vec-remove (vec %) index))
     (insert-removal element-route original-element removal-map)]))

(defn- replace-element [mast removal-map route newValue]
  (let [[mast removal-map] (remove-element mast removal-map route)]
    (add-element mast removal-map route newValue)))

(defn- apply-patch [in patch]
  (let [[mast removal-map] in
        {:keys [action route element value oldValue newValue name]} patch]
    (if (empty? route)
      [mast removal-map] ; Should not patch the root element at this time
      (case action
        "addElement" (add-element mast removal-map route element)
        "replaceElement" (replace-element mast removal-map route newValue)
        "addAttribute" (modify-element mast removal-map route element name nil value)
        "modifyAttribute" (modify-element mast removal-map route element name oldValue newValue)
        "removeAttribute" (modify-element mast removal-map route element name value nil)
        "removeElement" (remove-element mast removal-map route)))))

(defn- process-removal [path mast [index element]]
  (let [keyword-route path
        updated (update-in element [:labels] #(set (conj % :removed)))]
    (update-in mast keyword-route #(insert % index updated))))

(defn- process-removal-path [mast [path removals]]
  (reduce (partial process-removal path) mast removals))

(defn- apply-removal-map [mast removal-map]
  (let [mast-with-removals (reduce process-removal-path mast (reverse removal-map))]
    mast-with-removals))

(defn augment-mast-with-diff [mast diff]
  (let [removal-map {}
        [mast removal-map] (reduce apply-patch [mast removal-map] diff)]
    (apply-removal-map mast removal-map)))
