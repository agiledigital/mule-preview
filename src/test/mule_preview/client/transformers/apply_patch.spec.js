import { xml__GT_mast as xmlToMast } from "../../../../../dist/mule_preview.client.mast";
import { augment_mast_with_diff as augmentMastWithDiff } from "../../../../../dist/mule_preview.client.transformers.apply_patch";
import {
  xmlToClj,
  cljToJs,
  jsToClj
} from "../../../../../dist/mule_preview.client.test_utils";
import { readFileSync } from "fs";
import { join } from "path";
import exampleDiff from "./__fixtures__/example-diff-dom-output.json";

describe("when converting applying a diff to MAST", () => {
  it("transforms correctly", () => {
    const xmlFile = readFileSync(
      join(__dirname, "__fixtures__/example-mule-file.xml")
    );
    const clojureXmlMap = xmlToClj(xmlFile);
    const mast = xmlToMast(clojureXmlMap);
    const diff = jsToClj(exampleDiff);
    const transformedMast = cljToJs(augmentMastWithDiff(mast, diff));
    expect(transformedMast).toMatchSnapshot();
  });
});
