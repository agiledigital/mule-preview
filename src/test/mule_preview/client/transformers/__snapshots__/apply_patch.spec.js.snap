// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when converting applying a diff to MAST transforms correctly 1`] = `
Object {
  "attributes": Object {
    "xmlns": "http://www.mulesoft.org/schema/mule/core",
    "xmlns:db": "http://www.mulesoft.org/schema/mule/db",
    "xmlns:doc": "http://www.mulesoft.org/schema/mule/documentation",
    "xmlns:dw": "http://www.mulesoft.org/schema/mule/ee/dw",
    "xmlns:http": "http://www.mulesoft.org/schema/mule/http",
    "xmlns:mulexml": "http://www.mulesoft.org/schema/mule/xml",
    "xmlns:s3": "http://www.mulesoft.org/schema/mule/s3",
    "xmlns:spring": "http://www.springframework.org/schema/beans",
    "xmlns:tracking": "http://www.mulesoft.org/schema/mule/ee/tracking",
    "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
    "xsi:schemaLocation": "http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd
http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/http http://www.mulesoft.org/schema/mule/http/current/mule-http.xsd
http://www.mulesoft.org/schema/mule/db http://www.mulesoft.org/schema/mule/db/current/mule-db.xsd
http://www.mulesoft.org/schema/mule/ee/tracking http://www.mulesoft.org/schema/mule/ee/tracking/current/mule-tracking-ee.xsd
http://www.mulesoft.org/schema/mule/s3 http://www.mulesoft.org/schema/mule/s3/current/mule-s3.xsd
http://www.mulesoft.org/schema/mule/xml http://www.mulesoft.org/schema/mule/xml/current/mule-xml.xsd",
  },
  "content": Array [
    Object {
      "attributes": Object {
        "name": "example:/http-incoming-flow",
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": null,
              "description": "Overwrite the payload",
              "hash": 266487046,
              "labels": Array [
                "added",
              ],
              "location": null,
              "tag-name": "set-payload",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "config-ref": "example-pizza-httpListenerConfig",
                "doc:name": "HTTP",
                "path": "/",
              },
              "content-hash": -2017569654,
              "description": "HTTP",
              "hash": -699903627,
              "labels": Array [],
              "location": Object {
                "column": 95,
                "line": 14,
              },
              "tag-name": "http:listener",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Set Z level to 9000",
                "value": "#[9000]",
                "variableName": "z-level",
              },
              "change-record": Array [
                Object {
                  "delta": Array [
                    1907014272,
                    641237277,
                  ],
                  "name": "hash",
                },
                Object {
                  "delta": Array [
                    "Set Z level to 9000",
                    "Set Z level to 7000",
                  ],
                  "name": "description",
                },
              ],
              "content-hash": -2017569654,
              "description": "Set Z level to 7000",
              "hash": 641237277,
              "labels": Array [
                "edited",
              ],
              "location": Object {
                "column": 93,
                "line": 15,
              },
              "tag-name": "set-variable",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Overwrite the payload",
                "value": "#['lol']",
              },
              "content-hash": -2017569654,
              "description": "Overwrite the payload",
              "hash": 1820937142,
              "labels": Array [],
              "location": Object {
                "column": 72,
                "line": 16,
              },
              "tag-name": "set-payload",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "example:/expunge-convolvers",
                "name": "example:/expunge-convolvers",
              },
              "content-hash": -2017569654,
              "description": "example:/expunge-convolvers",
              "hash": 661336026,
              "labels": Array [],
              "location": Object {
                "column": 93,
                "line": 17,
              },
              "tag-name": "flow-ref",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Filter recalcitrant ions",
                "expression": "",
              },
              "content-hash": -2017569654,
              "description": "Filter recalcitrant ions",
              "hash": -586981530,
              "labels": Array [
                "removed",
              ],
              "location": Object {
                "column": 78,
                "line": 18,
              },
              "tag-name": "expression-filter",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "For Each",
              },
              "content": Array [
                Object {
                  "content": null,
                  "description": "example:/strain-overflow-noodle-requests",
                  "hash": -2070961420,
                  "labels": Array [
                    "added",
                  ],
                  "location": null,
                  "tag-name": "flow-ref",
                  "type": "component",
                },
                Object {
                  "content": null,
                  "description": "example:/strain-overflow-noodle-requests",
                  "hash": -2070961420,
                  "labels": Array [
                    "added",
                  ],
                  "location": null,
                  "tag-name": "flow-ref",
                  "type": "component",
                },
                Object {
                  "content": null,
                  "description": "example:/strain-overflow-noodle-requests",
                  "hash": -2070961420,
                  "labels": Array [
                    "added",
                  ],
                  "location": null,
                  "tag-name": "flow-ref",
                  "type": "component",
                },
                Object {
                  "content": null,
                  "description": "example:/strain-overflow-noodle-requests",
                  "hash": -2070961420,
                  "labels": Array [
                    "added",
                  ],
                  "location": null,
                  "tag-name": "flow-ref",
                  "type": "component",
                },
                Object {
                  "attributes": Object {
                    "doc:name": "example:/strain-overflow-noodle-requests",
                    "name": "example:/strain-overflow-noodle-requests",
                  },
                  "content-hash": -2017569654,
                  "description": "example:/strain-overflow-noodle-requests",
                  "hash": 160630889,
                  "labels": Array [],
                  "location": Object {
                    "column": 123,
                    "line": 20,
                  },
                  "tag-name": "flow-ref",
                  "type": "component",
                },
              ],
              "description": "For Each",
              "hash": 242079301,
              "labels": Array [
                "horizontal",
              ],
              "location": Object {
                "column": 37,
                "line": 19,
              },
              "tag-name": "foreach",
              "type": "container",
            },
            Object {
              "attributes": Object {
                "doc:name": "Server Noodles",
              },
              "content-hash": -2017569654,
              "description": "Server Noodles",
              "hash": -1584631798,
              "labels": Array [],
              "location": Object {
                "column": 48,
                "line": 22,
              },
              "tag-name": "set-payload",
              "type": "component",
            },
            Object {
              "attributes": Object {
                "doc:name": "Deduplicate",
              },
              "change-record": Array [
                Object {
                  "delta": Array [
                    385704797,
                    251600214,
                  ],
                  "name": "hash",
                },
              ],
              "content-hash": -539022716,
              "description": "Deduplicate",
              "hash": 251600214,
              "labels": Array [
                "edited",
              ],
              "location": Object {
                "column": 53,
                "line": 23,
              },
              "tag-name": "dw:transform-message",
              "type": "component",
            },
            Object {
              "content": null,
              "description": "Server Noodles 2",
              "hash": 798826535,
              "labels": Array [
                "added",
              ],
              "location": null,
              "tag-name": "set-payload",
              "type": "component",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "top",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
        Object {
          "content": Array [
            Object {
              "attributes": Object {
                "doc:name": "Catch Exception Strategy",
              },
              "content": Array [
                Object {
                  "content": null,
                  "description": "Nothing to see here",
                  "hash": 54737936,
                  "labels": Array [
                    "added",
                  ],
                  "location": null,
                  "tag-name": "set-payload",
                  "type": "component",
                },
                Object {
                  "attributes": Object {
                    "doc:name": "Nothing to see here",
                  },
                  "change-record": Array [
                    Object {
                      "delta": Array [
                        54737936,
                        266487046,
                      ],
                      "name": "hash",
                    },
                    Object {
                      "delta": Array [
                        "Nothing to see here",
                        "Overwrite the payload",
                      ],
                      "name": "description",
                    },
                  ],
                  "content-hash": -2017569654,
                  "description": "Overwrite the payload",
                  "hash": 266487046,
                  "labels": Array [
                    "edited",
                  ],
                  "location": Object {
                    "column": 57,
                    "line": 32,
                  },
                  "tag-name": "set-payload",
                  "type": "component",
                },
                Object {
                  "attributes": Object {
                    "doc:name": "Nor here",
                  },
                  "content-hash": -2017569654,
                  "description": "Nor here",
                  "hash": -1443655736,
                  "labels": Array [],
                  "location": Object {
                    "column": 65,
                    "line": 33,
                  },
                  "tag-name": "message-properties-transformer",
                  "type": "component",
                },
              ],
              "description": "Catch Exception Strategy",
              "hash": 497298084,
              "labels": Array [
                "error-handler",
              ],
              "location": Object {
                "column": 70,
                "line": 31,
              },
              "tag-name": "catch-exception-strategy",
              "type": "container",
            },
          ],
          "description": "",
          "labels": Array [
            "horizontal",
            "bottom",
          ],
          "tag-name": "psuedo",
          "type": "container",
        },
      ],
      "description": "example:/http-incoming-flow",
      "labels": Array [
        "vertical",
      ],
      "location": Object {
        "column": 45,
        "line": 13,
      },
      "tag-name": "flow",
      "type": "error-container",
    },
    Object {
      "attributes": Object {
        "name": "example:/expunge-convolvers",
      },
      "content": Array [
        Object {
          "attributes": Object {
            "bulkMode": "false",
            "config-ref": "",
            "doc:name": "Flush the convolver prisms",
          },
          "content-hash": -2017569654,
          "description": "Flush the convolver prisms",
          "hash": -1512076747,
          "labels": Array [],
          "location": Object {
            "column": 104,
            "line": 37,
          },
          "tag-name": "db:no-operation-selected",
          "type": "component",
        },
        Object {
          "attributes": Object {
            "doc:name": "Did the flush occur?",
          },
          "content": Array [
            Object {
              "attributes": Object {
                "expression": "",
              },
              "content": Array [
                Object {
                  "attributes": Object {
                    "config-ref": "",
                    "doc:name": "Fetch more prisms",
                  },
                  "content-hash": -2017569654,
                  "description": "Fetch more prisms",
                  "hash": -615782938,
                  "labels": Array [],
                  "location": Object {
                    "column": 86,
                    "line": 40,
                  },
                  "tag-name": "s3:no-operation-selected",
                  "type": "component",
                },
                Object {
                  "attributes": Object {
                    "doc:name": "Log number of fanglehockeys",
                    "level": "INFO",
                  },
                  "content-hash": -2017569654,
                  "description": "Log number of fanglehockeys",
                  "hash": 1560335807,
                  "labels": Array [],
                  "location": Object {
                    "column": 77,
                    "line": 41,
                  },
                  "tag-name": "logger",
                  "type": "component",
                },
              ],
              "description": null,
              "hash": 2014296490,
              "labels": Array [
                "horizontal",
              ],
              "location": Object {
                "column": 32,
                "line": 39,
              },
              "tag-name": "when",
              "type": "container",
            },
            Object {
              "attributes": Object {},
              "content": Array [
                Object {
                  "attributes": Object {
                    "doc:name": "Log Error",
                    "level": "INFO",
                    "message": "Bar",
                  },
                  "content-hash": -2017569654,
                  "description": "Log Error",
                  "hash": 2043302591,
                  "labels": Array [],
                  "location": Object {
                    "column": 73,
                    "line": 44,
                  },
                  "tag-name": "logger",
                  "type": "component",
                },
              ],
              "description": null,
              "hash": -15128758,
              "labels": Array [
                "horizontal",
              ],
              "location": Object {
                "column": 23,
                "line": 43,
              },
              "tag-name": "otherwise",
              "type": "container",
            },
          ],
          "description": "Did the flush occur?",
          "hash": -1105159260,
          "labels": Array [
            "vertical",
          ],
          "location": Object {
            "column": 48,
            "line": 38,
          },
          "tag-name": "choice",
          "type": "container",
        },
      ],
      "description": "example:/expunge-convolvers",
      "hash": 1066677628,
      "labels": Array [
        "horizontal",
        "removed",
      ],
      "location": Object {
        "column": 49,
        "line": 36,
      },
      "tag-name": "sub-flow",
      "type": "container",
    },
    Object {
      "attributes": Object {
        "name": "example:/strain-overflow-noodle-requests",
      },
      "content": Array [
        Object {
          "attributes": Object {
            "doc:name": "Clean carburetors",
          },
          "content": Array [
            Object {
              "attributes": Object {
                "doc:name": "Processor Chain",
              },
              "content": Array [
                Object {
                  "attributes": Object {
                    "config-ref": "",
                    "doc:name": "Clean Carburetor 1",
                    "method": "",
                    "path": "",
                  },
                  "content-hash": -2017569654,
                  "description": "Clean Carburetor 1",
                  "hash": 1805930470,
                  "labels": Array [],
                  "location": Object {
                    "column": 93,
                    "line": 51,
                  },
                  "tag-name": "http:request",
                  "type": "component",
                },
                Object {
                  "attributes": Object {
                    "doc:name": "XML to DOM",
                  },
                  "content-hash": -2017569654,
                  "description": "XML to DOM",
                  "hash": -269132085,
                  "labels": Array [],
                  "location": Object {
                    "column": 71,
                    "line": 52,
                  },
                  "tag-name": "mulexml:xml-to-dom-transformer",
                  "type": "component",
                },
              ],
              "description": "Processor Chain",
              "hash": 1690317879,
              "labels": Array [
                "horizontal",
              ],
              "location": Object {
                "column": 56,
                "line": 50,
              },
              "tag-name": "processor-chain",
              "type": "container",
            },
            Object {
              "attributes": Object {},
              "content": Array [
                Object {
                  "attributes": Object {
                    "config-ref": "",
                    "doc:name": "Oops this has a really long name",
                    "method": "",
                    "path": "",
                  },
                  "content-hash": -2017569654,
                  "description": "Oops this has a really long name",
                  "hash": 1450511924,
                  "labels": Array [],
                  "location": Object {
                    "column": 107,
                    "line": 55,
                  },
                  "tag-name": "http:request",
                  "type": "component",
                },
                Object {
                  "attributes": Object {
                    "doc:name": "Set Payload",
                  },
                  "content-hash": -2017569654,
                  "description": "Set Payload",
                  "hash": 2077915881,
                  "labels": Array [],
                  "location": Object {
                    "column": 53,
                    "line": 56,
                  },
                  "tag-name": "set-payload",
                  "type": "component",
                },
              ],
              "description": null,
              "hash": -15128758,
              "labels": Array [
                "horizontal",
              ],
              "location": Object {
                "column": 29,
                "line": 54,
              },
              "tag-name": "processor-chain",
              "type": "container",
            },
            Object {
              "attributes": Object {
                "config-ref": "",
                "doc:name": "Clean Carburetor 3",
                "method": "",
                "path": "",
              },
              "content-hash": -2017569654,
              "description": "Clean Carburetor 3",
              "hash": 1153502827,
              "labels": Array [],
              "location": Object {
                "column": 89,
                "line": 58,
              },
              "tag-name": "http:request",
              "type": "component",
            },
          ],
          "description": "Clean carburetors",
          "hash": -688832045,
          "labels": Array [
            "vertical",
          ],
          "location": Object {
            "column": 53,
            "line": 49,
          },
          "tag-name": "scatter-gather",
          "type": "container",
        },
        Object {
          "attributes": Object {
            "config-ref": "",
            "doc:name": "Start Internal Noodle Engine",
            "method": "",
            "path": "",
          },
          "content-hash": -2017569654,
          "description": "Start Internal Noodle Engine",
          "hash": -1535423867,
          "labels": Array [],
          "location": Object {
            "column": 95,
            "line": 60,
          },
          "tag-name": "http:request",
          "type": "component",
        },
      ],
      "description": "example:/strain-overflow-noodle-requests",
      "hash": 1816956808,
      "labels": Array [
        "horizontal",
      ],
      "location": Object {
        "column": 62,
        "line": 48,
      },
      "tag-name": "sub-flow",
      "type": "container",
    },
  ],
  "description": null,
  "hash": -83749566,
  "labels": Array [
    "vertical",
    "root",
  ],
  "location": Object {
    "column": 102,
    "line": 12,
  },
  "tag-name": "mule",
  "type": "container",
}
`;
